_episodes/16-parallel.md

```diff
@@ -25,632 +23,430 @@ We now have the tools we need to run a multi-processor job. This is a very
-{{ site.local.prompt }} python pi-serial.py 1000000
-Pi: 3.140944, memory: 0.022351741790771484 GiB
-{{ site.local.prompt }} python pi-serial.py 100000000
-Pi: 3.14182724, memory: 2.2351741790771484 GiB
-```
-{: .language-bash }
-
-Here we can see that the estimated amount of memory required scales linearly
-with the number of samples used.
-In practice, there is some memory required for other parts of the script,
-but the `x`, `y`, and `radii` variables are by far the largest influence
-on the total amount of memory required.
-
-### Estimating Calculation Time
+> The Amdahl program is written in Python, and installing or using it requires
+> locating the `python3` executable on the login node.
+> If it can't be found, try listing available modules using `module avail`,
+> load the appropriate one, and try the command again.
+{: .callout}
 
-Most of the calculations required to estimate π are in the
-`inside_circle` function:
+### MPI for Python
 
-1. Generating `n_samples` random values for `x` and `y`.
-1. Calculating `n_samples` values of `radii` from `x` and `y`.
-1. Counting how many values in `radii` are under 1.0.
+The Amdahl code has one dependency: __mpi4py__.
+If it hasn't already been installed on the cluster, `pip` will attempt to
+collect mpi4py from the Internet and install it for you.
+If this fails due to a one-way firewall, you must retrieve mpi4py on your
+local machine and upload it, just as we did for Amdahl.
 
-There's also one multiplication operation and one division operation required
-to convert the `counts` value to the final estimate of π in the main
-function.
+> ## Retrieve and Push mpi4py
+>
+> If installing Amdahl failed because mpi4py could not be installed,
+> retrieve the tarball from <https://github.com/mpi4py/mpi4py/tarball/master>
+> then `rsync` it to the cluster, extract, and install:
+>
+> ```
+> {{ site.local.prompt }} wget https://github.com/mpi4py/mpi4py/tarball/master -O mpi4py.tar.gz
+> {{ site.local.prompt }} scp amdahl.tar.gz {{ site.remote.user }}@{{ site.remote.login }}:
+> # or
+> {{ site.local.prompt }} rsync -avP amdahl.tar.gz {{ site.remote.user }}@{{ site.remote.login }}:
+> ```
+> {: .language-bash}
+> ```
+> {{ site.remote.prompt }} mkdir mpi4py
+> {{ site.remote.prompt }} tar -xvzf mpi4py.tar.gz -C amdahl --strip-components=1
+> {{ site.remote.prompt }} cd mpi4py
+> {{ site.remote.prompt }} python3 -m pip install --user .
+> {{ site.remote.prompt }} cd ../amdahl
+> {{ site.remote.prompt }} python3 -m pip install --user .
+> ```
+> {: .language-bash}
+{: .discussion}
 
-A simple way to measure the calculation time is to use Python's `datetime`
-module to store the computer's current date and time before and after the
-calculations, and calculate the difference between those times.
+`pip` may warn that your user package binaries are not in your PATH.
 
-To add the time measurement to the script, add the following line below the
-`import sys` line:
+> ## If Pip Raises a Warning...
```

Comment:
```suggestion
> ## If `pip` Raises a Warning...
```

Reply:
Pip is the proper noun, `pip` is the command (or, more recently, the Python module to invoke). The heading is "correct" as written, but if it causes confusion or increases cognitive loading, I'm open to changing it.